src/detect-xbits.c

```diff
@@ -305,25 +215,26 @@ static int DetectXbitParse(DetectEngineCtx *de_ctx,
         case DETECT_XBITS_CMD_UNSET:
         case DETECT_XBITS_CMD_TOGGLE:
         default:
-            if (strlen(fb_name) == 0)
+            if (strlen(ptr->name) == 0) {
+                rs_xbits_free(ptr, ptr->name);
                 return -1;
+            }
             break;
     }
 
     cd = SCMalloc(sizeof(DetectXbitsData));
     if (unlikely(cd == NULL))
         return -1;
 
-    cd->idx = VarNameStoreSetupAdd(fb_name, var_type);
-    cd->cmd = fb_cmd;
-    cd->tracker = hb_dir;
-    cd->type = var_type;
-    cd->expire = expire;
-
-    SCLogDebug("idx %" PRIu32 ", cmd %s, name %s",
-        cd->idx, fb_cmd_str, strlen(fb_name) ? fb_name : "(none)");
+    cd->idx = VarNameStoreSetupAdd(ptr->name, ptr->vartype);
+    cd->cmd = ptr->cmd;
+    cd->tracker = ptr->tracker;
+    cd->type = ptr->vartype;
+    cd->expire = ptr->expire;
 
     *cdout = cd;
+    // ptr's job is done, free it
+    rs_xbits_free(ptr, ptr->name);
```

Comment:
Should this only be freed once in the `free` function for the keyword?

Reply:
could the Rust side just completely own `DetectXbitsData` instead of how it is here, where there is a mix?

Reply:
Indeed. I had thought about it but was not parsing and passing all the params back then and it is used at more places than just `detect-xbits`. I think it is possible with the current struct though.